janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/backpressure/SemaphoreQueryBackPressure.java

```diff
@@ -0,0 +1,65 @@
+// Copyright 2023 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.diskstorage.util.backpressure;
+
+import org.janusgraph.core.JanusGraphException;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+
+import static org.janusgraph.util.system.ExecuteUtil.gracefulExecutorServiceShutdown;
+
+public class SemaphoreQueryBackPressure implements QueryBackPressure{
+
+    private final ExecutorService executorService = Executors.newSingleThreadExecutor();
+    private final Runnable releaseNonBlocking;
+    private final Semaphore semaphore;
+
+    public SemaphoreQueryBackPressure(final int backPressureLimit) {
+        this.semaphore = new Semaphore(backPressureLimit, true);
+        this.releaseNonBlocking = () -> {
+            // ensure we never add more permits than `backPressureLimit`
+            // (even if `releaseAfterQuery()` is called more times than `acquireBeforeQuery()`);
+            if(semaphore.availablePermits()<backPressureLimit){
+                semaphore.release();
+            }
+        };
+    }
+
+    @Override
+    public void acquireBeforeQuery() {
+        try {
+            semaphore.acquire();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new JanusGraphException(e);
+        }
+    }
+
+    @Override
+    public void releaseAfterQuery(){
+        executorService.execute(releaseNonBlocking);
```

Comment:
It’s a good question.
The idea was that we need to perform a locking operation here. I.e. we need to do something if something else matches. 
‘if(semaphore.availablePermits()<backPressureLimit) {}’ - this check could be executed concurrently which may signal multiple threads that they can execute ‘release’ operation. In this situation we may accidentally release more permits than necessary.
Logically we need to lock this check and the next operation (synchronise it), so that only a single thread could  execute that critical section at any given time.
The problem with locking (synchronisation) os that it’s a blocking operation (sometimes), but this release is executed at CQL IO thread. It’s a bad idea to block CQL IO thread for anything. Moreover the processing thread isn’t actually interested in awaiting for the result. It only wants to signal that the release must happen, but doesn’t interested to wait until it is released. 
In such case, it’s better to use a single dedicated thread here (always a single thread and nothing else) to process releases because we can avoid synchronisation here (as the critical section is always accessed by one thread, thus not needed to be synchronised), plus in situations when multiple threads are accessing the same critical section - the call of this method won’t be blocked.